master/internal/resourcemanagers/agent/agent.go

```diff
@@ -92,16 +95,35 @@ func (a *agent) receive(ctx *actor.Context, msg interface{}) error {
 		socket, ok := msg.Accept(ctx, aproto.MasterMessage{}, true)
 		check.Panic(check.True(ok, "failed to accept websocket connection"))
 		a.socket = socket
+
 		lastColonIndex := strings.LastIndex(msg.Ctx.Request().RemoteAddr, ":")
 		if lastColonIndex == -1 {
 			a.address = msg.Ctx.Request().RemoteAddr
 		} else {
 			a.address = msg.Ctx.Request().RemoteAddr[0:lastColonIndex]
 		}
-		wsm := ws.WriteMessage{Message: aproto.AgentMessage{MasterSetAgentOptions: a.opts}}
+
+		var masterSetAgentOptions aproto.AgentMessage
+		if a.awaitingReconnect && a.agentReattachEnabled {
+			optsCopy := *a.opts
+			optsCopy.ContainersToReattach = a.gatherContainersToReattach(ctx)
+			masterSetAgentOptions = aproto.AgentMessage{MasterSetAgentOptions: &optsCopy}
+		} else {
```

Comment:
excuse me for thinking out loud here a bit.

same comment here about enumerating cases explicitly; here it is awaitingReconnect x agentReattachEnabled x IsReconnect, which is a lot of cases, sure, but 'compressing' them requires the reader to carefully trace the code to check we handle each given case correctly. tracing agentReattachEnabled == false, IsReconnect == true, awaitingReconnect == false in isolation, i see we may have an agent reconnect (maybe with containers still active), but since we think the agent is reconnecting cleanly we don't clear its containers, but awaitingReconnect == false implies we may've forgotten about its containers. but of course IsReconnect should always imply awaitingReconnect when the actor is still alive, and if the agent actor wasn't alive, we wouldn't reconnect it improperly but kick the agent away in the agents actor while it had IsReconnect == true (and if was reconnecting while "still connected", we always panic on a second connecting), so this case is bogus (not reachable). but we handle it as if it is possible and i think that is what makes this hard to grok (that being said, if it _were_possible, killing the containers would make sense I think). i think verifying this is correct would be a lot easier with them all enumerated with comments, and we called out cases like this. of course, explicitly specifying the dependent variables and verifying invariants would help too (like having the invariant IsReconnect implies awaitingReconnect explicit, and explaining how it is upheld, in the code would be helpful).

in general i've thought about this some before, when i was trying to write the trial actor and trying to rewrite the trial in a way i knew was correct, and there are a spectrum of options between two ends: enumerating cases explicitly, entirely, or fulling decomposing the case space into regions that express like cases with if/else and boolean logic (taking advantage of implicit relationships between variables). the former is wordy, and you have to use routines to not repeat yourself, but has the advantage of allowing less mistakes since all cases are thought through by force, the latter is the norm (as a side effect how we all write code) but does allow for subtle mistakes by nature the added work of attempting to simplify/specify a function and making assumption about invariants that may not be upheld.

so i guess my takeaway here is maybe this is complicated enough and the cases are subtle enough that enumerating more of them explicitly, panicking on invalid states and leaving good comments about intended behavior could be fun. idk... but it's been fun to think about.